File: android/app/src/main/java/top/ourfor/app/iPlayClient/view/PlayerControlView.java
package top.ourfor.app.iPlayClient.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.view.View;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.constraintlayout.widget.ConstraintLayout;

import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.atomic.AtomicInteger;

import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import top.ourfor.app.iPlayClient.R;
import top.ourfor.app.iPlayClient.helper.IntervalCaller;

@RequiresApi(api = Build.VERSION_CODES.O)
@Setter
@Getter
@Slf4j
public class PlayerControlView extends ConstraintLayout implements PlayerEventListener {
    public static final int ICON_SMALL_SIZE = 24 * 3;
    public static final int ICON_SIZE = 32 * 3;
    public static final int ICON_TAG = 2;
    public PlayerEventListener delegate;
    private boolean shouldUpdateProgress = true;
    public Player player;
    private String videoTitle;
    private AtomicInteger resId = new AtomicInteger(8000);
    private DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("HH:mm:ss");
    private IntervalCaller updateProgressCaller = new IntervalCaller(1000, 0);

    public PlayerControlView(Context context) {
        super(context);
        setupUI();
        bind();
    }

    public void setVideoTitle(String value) {
        videoTitle = value;
        titleLabel.setText(value);
    }

    public void updateFullscreenStyle(boolean isFullscreen) {
        if (isFullscreen) {
            updateIcon(fullscreenButton, R.drawable.arrow_up_right_and_arrow_down_left);
        } else {
            updateIcon(fullscreenButton, R.drawable.viewfinder);
        }
    }

    public View playButton = new ConstraintLayout(getContext()) {
        {
            ConstraintLayout layout = this;
            ImageView icon = new ImageView(getContext());
            icon.setTag(ICON_TAG);
            icon.setImageResource(R.drawable.pause);
            icon.setImageTintList(ColorStateList.valueOf(Color.WHITE));
            LayoutParams iconLayout = centerLayout();
            iconLayout.width = ICON_SIZE;
            iconLayout.height = ICON_SIZE;
            layout.addView(icon, iconLayout);
            GradientDrawable gradientDrawable = new GradientDrawable();
            gradientDrawable.setColor(Color.argb(50, 0, 0, 0));
            gradientDrawable.setCornerRadius(ICON_SIZE + 0f);
            layout.setBackground(gradientDrawable);
            layout.getRootView().setId(resId.getAndIncrement());
            layout.getRootView();
        }
    };

    private LayoutParams playButtonLayout = new LayoutParams(ICON_SIZE * 2, ICON_SIZE * 2) {
        {
            topToTop = LayoutParams.PARENT_ID;
            bottomToBottom = LayoutParams.PARENT_ID;
            rightToRight = LayoutParams.PARENT_ID;
            leftToLeft = LayoutParams.PARENT_ID;
        }
    };

    public View fullscreenButton = new ConstraintLayout(getContext()) {
        {
            ConstraintLayout layout = this;
            ImageView icon = new ImageView(getContext());
            icon.setTag(ICON_TAG);
            icon.setImageResource(R.drawable.viewfinder);
            icon.setImageTintList(ColorStateList.valueOf(Color.WHITE));
            LayoutParams iconLayout = new LayoutParams(centerLayout());
            iconLayout.width = ICON_SMALL_SIZE;
            iconLayout.height = ICON_SMALL_SIZE;
            layout.addView(icon, iconLayout);
            GradientDrawable gradientDrawable = new GradientDrawable();
            gradientDrawable.setColor(Color.argb(50, 0, 0, 0));
            gradientDrawable.setCornerRadius(ICON_SMALL_SIZE + 0f);
            layout.setBackground(gradientDrawable);
            layout.getRootView().setId(resId.getAndIncrement());
            layout.getRootView();
        }
    };

    private LayoutParams fullscreenLayout = new LayoutParams(ICON_SMALL_SIZE * 2, ICON_SMALL_SIZE * 2) {
        {
            topToTop = LayoutParams.PARENT_ID;
            rightToRight = LayoutParams.PARENT_ID;
            topMargin = 48;
            rightMargin = 48;
        }
    };

    public View subtitleButton = new ConstraintLayout(getContext()) {
        {
            ConstraintLayout layout = this;
            ImageView icon = new ImageView(getContext());
            icon.setTag(ICON_TAG);
            icon.setImageResource(R.drawable.captions_bubble);
            icon.setImageTintList(ColorStateList.valueOf(Color.WHITE));
            LayoutParams iconLayout = new LayoutParams(centerLayout());
            iconLayout.width = ICON_SMALL_SIZE;
            iconLayout.height = ICON_SMALL_SIZE;
            layout.addView(icon, iconLayout);
            GradientDrawable gradientDrawable = new GradientDrawable();
            gradientDrawable.setColor(Color.argb(50, 0, 0, 0));
            gradientDrawable.setCornerRadius(ICON_SMALL_SIZE + 0f);
            layout.setBackground(gradientDrawable);
            layout.getRootView().setId(resId.getAndIncrement());
            layout.getRootView();
        }
    };

    private LayoutParams subtitleButtonLayout = new LayoutParams(ICON_SMALL_SIZE * 2, ICON_SMALL_SIZE * 2) {
        {
            topToTop = LayoutParams.PARENT_ID;
            rightToLeft = fullscreenButton.getId();
            topMargin = 48;
            rightMargin = 48;
        }
    };

    public View audioButton = new ConstraintLayout(getContext()) {
        {
            ConstraintLayout layout = this;
            ImageView icon = new ImageView(getContext());
            icon.setTag(ICON_TAG);
            icon.setImageResource(R.drawable.waveform);
            icon.setImageTintList(ColorStateList.valueOf(Color.WHITE));
            LayoutParams iconLayout = new LayoutParams(centerLayout());
            iconLayout.width = ICON_SMALL_SIZE;
            iconLayout.height = ICON_SMALL_SIZE;
            layout.addView(icon, iconLayout);
            GradientDrawable gradientDrawable = new GradientDrawable();
            gradientDrawable.setColor(Color.argb(50, 0, 0, 0));
            gradientDrawable.setCornerRadius(ICON_SMALL_SIZE + 0f);
            layout.setBackground(gradientDrawable);
            layout.getRootView().setId(resId.getAndIncrement());
            layout.getRootView();
        }
    };

    private LayoutParams audioButtonLayout = new LayoutParams(ICON_SMALL_SIZE * 2, ICON_SMALL_SIZE * 2) {
        {
            topToTop = LayoutParams.PARENT_ID;
            rightToLeft = subtitleButton.getId();
            topMargin = 48;
            rightMargin = 48;
        }
    };

    public PlayerSlider progressBar = new PlayerSlider(getContext()) {
        {
            int color = Color.WHITE;
            ColorStateList colorStateList = ColorStateList.valueOf(color);
            int thumbRadius = 25;
            GradientDrawable thumb = new GradientDrawable();
            thumb.setShape(GradientDrawable.OVAL);
            thumb.setSize(thumbRadius * 2, thumbRadius * 2);
            thumb.setColor(Color.RED);
            setThumb(thumb);
            setThumbTintList(colorStateList);
            setProgressTintList(colorStateList);
            setId(resId.getAndIncrement());
            setPadding(thumbRadius, thumbRadius, thumbRadius, thumbRadius);
        }
    };


    private LayoutParams progressBarLayout = new LayoutParams(0, 60) {
        {
            matchConstraintPercentWidth = 0.9f;
            bottomToBottom = LayoutParams.PARENT_ID;
            bottomMargin = 100;
            leftToLeft = LayoutParams.PARENT_ID;
            rightToRight = LayoutParams.PARENT_ID;
        }
    };

    @SuppressLint("AppCompatCustomView")
    private TextView durationLabel = new TextView(getContext()) {
        {
            setTextSize(12.0F);
            setTextColor(Color.WHITE);
            setText(formatTime(0, 0));
            setId(resId.getAndIncrement());
        }
    };

    private LayoutParams durationLayout = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT) {
        {
            rightToRight = progressBar.getId();
            bottomToTop = progressBar.getId();
            bottomMargin = 10;
        }
    };

    @SuppressLint("AppCompatCustomView")
    public TextView titleLabel = new TextView(getContext()) {
        {
            setTextSize(14.0F);
            setTextColor(Color.WHITE);
            setId(resId.getAndIncrement());
        }
    };

    private LayoutParams titleLayout = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT) {
        {
            leftToLeft = progressBar.getId();
            bottomToTop = progressBar.getId();
            matchConstraintPercentWidth = 0.75f;
            bottomMargin = 10;
        }
    };

    private void setupUI() {
        addView(playButton, playButtonLayout);
        addView(fullscreenButton, fullscreenLayout);
        addView(subtitleButton, subtitleButtonLayout);
        addView(audioButton, audioButtonLayout);
        addView(progressBar, progressBarLayout);
        addView(durationLabel, durationLayout);
        addView(titleLabel, titleLayout);
    }

    private void bind() {
        playButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                log.debug("play");
                boolean isPlaying = player != null && player.isPlaying();
                int resId = isPlaying ? R.drawable.play : R.drawable.pause;
                post(() -> updateIcon(playButton, resId));
                if (isPlaying) {
                    player.pause();
                } else {
                    player.resume();
                }
            }
        });
        progressBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                shouldUpdateProgress = false;
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                player.seek(progressBar.getProgress());
                shouldUpdateProgress = true;
            }
        });
        fullscreenButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                delegate.onWindowSizeChange();
            }
        });

        subtitleButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                delegate.onSelectSubtitle();
            }
        });

        audioButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                delegate.onSelectAudio();
            }
        });
    }

    public void toggleVisible() {
        float newAlpha = 1.0f;
        if (getAlpha() == 1.0f) {
            newAlpha = 0.0f;
        }
        animate()
                .alpha(newAlpha)
                .setDuration(800)
                .start();
    }

    public void updateControlVisible(boolean visible) {
        float newAlpha = 0.0f;
        if (visible) {
            newAlpha = 1.0f;
        }
        if (newAlpha == getAlpha()) {
            return;
        }

        animate()
                .alpha(newAlpha)
                .setDuration(800)
                .start();
    }

    private void updateIcon(View view, int resId) {
        ImageView imageView = view.findViewWithTag(ICON_TAG);
        if (!(imageView instanceof ImageView)) {
            return;
        }
        imageView.setImageResource(resId);
        imageView.setImageTintList(ColorStateList.valueOf(Color.WHITE));
    }

    @Override
    public void onPropertyChange(PlayerPropertyType name, Object value) {
        if (value == null) {
            return;
        }
        if (name == PlayerPropertyType.Duration) {
            double duration = (double) value;
            progressBar.setMax((int) duration);
            durationLabel.setText(Double.toString(duration));
            durationLabel.setText(formatTime(progressBar.getProgress(), progressBar.getMax()));
        } else if (name == PlayerPropertyType.TimePos) {
            if (!shouldUpdateProgress) {
                return;
            }

            double time = (double) value;
            updateProgressCaller.invoke(() -> post(() -> {
                progressBar.setProgress((int) time);
                durationLabel.setText(formatTime(progressBar.getProgress(), progressBar.getMax()));
            }));
        } else if (name == PlayerPropertyType.EofReached) {
            boolean isEof = (boolean)value;
            if (isEof) {
                updateIcon(playButton, R.drawable.play);
            }
        }
    }

    private String formatTime(int current, int total) {
        Duration duration = Duration.ofSeconds(current);
        LocalDateTime time = LocalDateTime.MIN.plus(duration);
        String part1 = time.format(dateFormatter);
        duration = Duration.ofSeconds(total);
        time = LocalDateTime.MIN.plus(duration);
        String part2 = time.format(dateFormatter);
        return part1 + " / " + part2;
    }

    public static LayoutParams centerLayout() {
        LayoutParams centerParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        centerParams.topToTop = LayoutParams.PARENT_ID;
        centerParams.leftToLeft = LayoutParams.PARENT_ID;
        centerParams.rightToRight = LayoutParams.PARENT_ID;
        centerParams.bottomToBottom = LayoutParams.PARENT_ID;
        return centerParams;
    }
}

File: android/app/src/main/java/top/ourfor/app/iPlayClient/PlayerControlView.kt
package top.ourfor.app.iPlayClient

import android.animation.ValueAnimator
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.constraintlayout.widget.ConstraintLayout
import java.time.Duration
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.atomic.AtomicInteger


@RequiresApi(Build.VERSION_CODES.O)
class PlayerControlView(context: Context) : ConstraintLayout(context), PlayerEventListener {
    public var delegate: PlayerEventListener? = null;
    private var shouldUpdateProgress = true;
    var player: Player? = null
    var videoTitle: String? = null
        set(value) {
            field = value
            titleLabel.text = value
        }

    fun updateFullscreenStyle(isFullscreen: Boolean) {
        if (isFullscreen) {
            updateIcon(fullscreenButton, androidx.media3.ui.R.drawable.exo_icon_fullscreen_exit)
        } else {
            updateIcon(fullscreenButton, androidx.media3.ui.R.drawable.exo_icon_fullscreen_enter)
        }
    }

    private var resId: AtomicInteger = AtomicInteger(8000)
    private val dateFormatter: DateTimeFormatter by lazy {
        DateTimeFormatter.ofPattern("HH:mm:ss")
    }

    var playButton: View = run {
        val layout = ConstraintLayout(context)
        val icon = ImageView(context)
        icon.tag = ICON_TAG
        icon.setImageResource(androidx.media3.ui.R.drawable.exo_icon_pause)
        val iconLayout = centerLayout()
        iconLayout.width = ICON_SIZE
        iconLayout.height = ICON_SIZE
        layout.addView(icon, iconLayout)
        val gradientDrawable = GradientDrawable()
        gradientDrawable.setColor(Color.argb(50, 0, 0, 0))
        gradientDrawable.cornerRadius = ICON_SIZE + 0f
        layout.background = gradientDrawable
        layout.rootView.id = resId.getAndIncrement()
        layout.rootView
    }

    private var playButtonLayout = run {
        val params = LayoutParams(ICON_SIZE * 2, ICON_SIZE * 2)
        params.topToTop = LayoutParams.PARENT_ID
        params.bottomToBottom = LayoutParams.PARENT_ID
        params.rightToRight = LayoutParams.PARENT_ID
        params.leftToLeft = LayoutParams.PARENT_ID
        params
    }

    var fullscreenButton = run {
        val layout = ConstraintLayout(context)
        val icon = ImageView(context)
        icon.tag = ICON_TAG
        icon.setImageResource(androidx.media3.ui.R.drawable.exo_icon_fullscreen_enter)
        val iconLayout = LayoutParams(centerLayout())
        iconLayout.width = ICON_SMALL_SIZE
        iconLayout.height = ICON_SMALL_SIZE
        layout.addView(icon, iconLayout)
        val gradientDrawable = GradientDrawable()
        gradientDrawable.setColor(Color.argb(50, 0, 0, 0))
        gradientDrawable.cornerRadius = ICON_SMALL_SIZE + 0f
        layout.background = gradientDrawable
        layout.rootView.id = resId.getAndIncrement()
        layout.rootView
    }

    private var fullscreenLayout = run {
        val params = LayoutParams(ICON_SMALL_SIZE * 2, ICON_SMALL_SIZE * 2)
        params.topToTop = LayoutParams.PARENT_ID
        params.rightToRight = LayoutParams.PARENT_ID
        params.topMargin = 48
        params.rightMargin = 48
        params
    }

    var progressBar = run {
        val slide = SeekBar(context)
        val color = Color.WHITE
        val colorStateList = ColorStateList.valueOf(color)
        val thumbRadius = 25
        val thumb = GradientDrawable()
        thumb.shape = GradientDrawable.OVAL
        thumb.setSize(thumbRadius * 2, thumbRadius * 2)
        thumb.setColor(Color.RED)
        slide.thumb = thumb
        slide.thumbTintList = colorStateList
        slide.progressTintList = colorStateList
        slide.id = resId.getAndIncrement()
        slide.setPadding(thumbRadius, thumbRadius, thumbRadius, thumbRadius)
        slide
    }

    private var progressBarLayout = run {
        val params = LayoutParams(0, 60)
        params.matchConstraintPercentWidth = 0.9f
        params.bottomToBottom = LayoutParams.PARENT_ID
        params.bottomMargin = 100
        params.leftToLeft = LayoutParams.PARENT_ID
        params.rightToRight = LayoutParams.PARENT_ID
        params
    }

    private var durationLabel = run {
        val label = TextView(context)
        label.textSize = 12.0F
        label.setTextColor(Color.WHITE)
        label.text = formatTime(0, 0)
        label.id = resId.getAndIncrement()
        label
    }

    private var durationLayout = run {
        val params = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        params.rightToRight = progressBar.id
        params.bottomToTop = progressBar.id
        params.bottomMargin = 10
        params
    }

    var titleLabel = run {
        val label = TextView(context)
        label.textSize = 14.0F
        label.setTextColor(Color.WHITE)
        label.id = resId.getAndIncrement()
        label
    }

    private var titleLayout = run {
        val params = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        params.leftToLeft = progressBar.id
        params.bottomToTop = progressBar.id
        params.bottomMargin = 10
        params
    }

    var numberValueView = run {
        val view = PlayerNumberValueView(context);
        view.id = resId.getAndIncrement()
        view
    }

    private var numberValueLayout = run {
        val params = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        params.leftToLeft = LayoutParams.PARENT_ID;
        params.topToTop = LayoutParams.PARENT_ID;
        params.rightToRight = LayoutParams.PARENT_ID;
        params.topMargin = 100
        params
    }


    init {
        setupUI()
        bind()
    }

    private fun setupUI() {
        addView(playButton, playButtonLayout)
        addView(fullscreenButton, fullscreenLayout)
        addView(progressBar, progressBarLayout)
        addView(durationLabel, durationLayout)
        addView(titleLabel, titleLayout)
        addView(numberValueView, numberValueLayout)
    }

    private fun bind() {
        playButton.setOnClickListener {
            Log.d(TAG, "play")
            val isPlaying = player?.isPlaying == true
            var resId = if (isPlaying) androidx.media3.ui.R.drawable.exo_icon_play else androidx.media3.ui.R.drawable.exo_icon_pause
            post {
                updateIcon(playButton, resId)
            }
            if (isPlaying) {
                player?.pause()
            } else {
                player?.resume()
            }
        }
        progressBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                shouldUpdateProgress = false;
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                player?.seek(progressBar.progress.toLong())
                shouldUpdateProgress = true;
            }
        })
        fullscreenButton.setOnClickListener {
            delegate?.onWindowSizeChange()
        }
    }

    fun toggleVisible() {
        animate()
            .alpha(1.0f - alpha)
            .setDuration(800)
            .start()
    }

    private fun updateIcon(view: View, resId: Int) {
        val imageView = view.findViewWithTag<ImageView?>(ICON_TAG)
        if (imageView !is ImageView) return
        imageView?.setImageResource(resId)
    }

    override fun onPropertyChange(name: String?, value: Any?) {
        if (value == null) return
        if (name.equals("duration")) {
            val duration = value as Double
            progressBar.max = duration.toInt()
            durationLabel.text = duration.toString()
            durationLabel.text = formatTime(progressBar.progress, progressBar.max)
            requestLayout()
        } else if (name.equals("time-pos")) {
            if (!shouldUpdateProgress) {
                return
            }

            val time = value as Double
            progressBar.progress = time.toInt()
            durationLabel.text = formatTime(progressBar.progress, progressBar.max)
        }
    }

    private fun formatTime(current: Int, total: Int): String {
        var duration: Duration = Duration.ofSeconds(current.toLong())
        var time = LocalDateTime.MIN.plus(duration)
        val part1 = time.format(dateFormatter)
        duration = Duration.ofSeconds(total.toLong())
        time = LocalDateTime.MIN.plus(duration)
        val part2 = time.format(dateFormatter)
        return "$part1 / $part2"
    }

    companion object {
        val TAG = "PlayerControlView"

        fun centerLayout(): LayoutParams {
            val centerParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
            centerParams.topToTop = LayoutParams.PARENT_ID;
            centerParams.leftToLeft = LayoutParams.PARENT_ID
            centerParams.rightToRight = LayoutParams.PARENT_ID
            centerParams.bottomToBottom = LayoutParams.PARENT_ID
            return centerParams
        }

        val ICON_SMALL_SIZE = 24 * 3
        val ICON_SIZE = 32 * 3
        val ICON_TAG = 2
    }
}